packages/gator-permissions-snap/src/services/snapsMetricsService.ts

```diff
@@ -0,0 +1,245 @@
+/* eslint-disable @typescript-eslint/naming-convention */
+import { logger } from '@metamask/7715-permissions-shared/utils';
+import type { Json, SnapsProvider } from '@metamask/snaps-sdk';
+
+/**
+ * Permission value details for analytics.
+ */
+export type PermissionValue = {
+  /** The period in seconds (for periodic permissions) */
+  period?: number;
+  /** The amount in the token's base unit */
+  amount?: string;
+  /** The token address (0x0 for native tokens) */
+  token?: string;
+  /** The chain ID */
+  chainId?: string;
+  /** Duration/expiry in seconds */
+  duration?: number;
+};
+
+/**
+ * Service for tracking Snaps metrics events in the snap.
+ * Uses the snap_trackEvent API to send events to MetaMask.
+ */
+export class SnapsMetricsService {
+  readonly #snap: SnapsProvider;
+
+  constructor(snap: SnapsProvider) {
+    this.#snap = snap;
+  }
+
+  /**
+   * Track a generic event with custom properties.
+   * @param event - The event name.
+   * @param properties - Additional properties to include with the event.
+   */
+  async #trackEvent(
+    event: string,
+    properties: Record<string, Json>,
+  ): Promise<void> {
+    try {
+      logger.debug(`SnapsMetricsService: Tracking event ${event}`, properties);
+      await this.#snap.request({
+        method: 'snap_trackEvent',
+        params: {
+          event: {
+            event: `${event}`,
+            properties,
+          },
+        },
+      });
+    } catch (error) {
+      logger.error(`SnapsMetricsService: Failed to track event ${event}`, {
+        error,
+        event,
+        properties,
+      });
+    }
+  }
+
+  /**
+   * Track when a permission request is initiated.
+   * @param origin - The origin of the request.
+   * @param permissionType - The type of permission being requested.
+   * @param permissionValue - The permission value details.
+   */
+  async trackPermissionRequestStarted(
+    origin: string,
+    permissionType: string,
+    permissionValue?: PermissionValue,
```

Comment:
This comment goes for all of the track*** functions - it would be awesome to make these named parameters, rather than positional. Especially when you end up with multiple primitive values, it's difficult to understand what they are, and high risk to mis-align them.

Because all of the track** functions have the same base `{ origin: string, permissionType: string }` this could be defined in a shared type.

Reply:
Good point. Agree that using named parameters would make the track functions much clearer and less error-prone.